tokio/tests/task_id.rs

```diff
@@ -0,0 +1,298 @@
+#![warn(rust_2018_idioms)]
+#![allow(clippy::declare_interior_mutable_const)]
+#![cfg(all(feature = "full", not(tokio_wasi)))]
+
+#[cfg(tokio_unstable)]
+use std::error::Error;
+#[cfg(tokio_unstable)]
+use std::future::Future;
+#[cfg(tokio_unstable)]
+use std::pin::Pin;
+#[cfg(tokio_unstable)]
+use std::task::{Context, Poll};
+#[cfg(tokio_unstable)]
+use tokio::runtime::{Builder, Runtime};
+#[cfg(tokio_unstable)]
+use tokio::sync::oneshot;
+#[cfg(tokio_unstable)]
+use tokio::task::{self, LocalSet};
+
+#[cfg(tokio_unstable)]
+mod support {
+    pub mod panic;
+}
+#[cfg(tokio_unstable)]
+use support::panic::test_panic;
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "current_thread")]
+async fn task_id_spawn() {
+    tokio::spawn(async { println!("task id: {}", task::id()) })
+        .await
+        .unwrap();
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "current_thread")]
+async fn task_id_spawn_blocking() {
+    task::spawn_blocking(|| println!("task id: {}", task::id()))
+        .await
+        .unwrap();
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "current_thread")]
+async fn task_id_collision_current_thread() {
+    let handle1 = tokio::spawn(async { task::id() });
+    let handle2 = tokio::spawn(async { task::id() });
+
+    let (id1, id2) = tokio::join!(handle1, handle2);
+    assert_ne!(id1.unwrap(), id2.unwrap());
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "multi_thread")]
+async fn task_id_collision_multi_thread() {
+    let handle1 = tokio::spawn(async { task::id() });
+    let handle2 = tokio::spawn(async { task::id() });
+
+    let (id1, id2) = tokio::join!(handle1, handle2);
+    assert_ne!(id1.unwrap(), id2.unwrap());
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "current_thread")]
+async fn task_ids_match_current_thread() {
+    let (tx, rx) = oneshot::channel();
+    let handle = tokio::spawn(async {
+        let id = rx.await.unwrap();
+        assert_eq!(id, task::id());
+    });
+    tx.send(handle.id()).unwrap();
+    handle.await.unwrap();
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "multi_thread")]
+async fn task_ids_match_multi_thread() {
+    let (tx, rx) = oneshot::channel();
+    let handle = tokio::spawn(async {
+        let id = rx.await.unwrap();
+        assert_eq!(id, task::id());
+    });
+    tx.send(handle.id()).unwrap();
+    handle.await.unwrap();
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "multi_thread")]
+async fn task_id_future_destructor_completion() {
+    struct MyFuture;
+
+    impl Future for MyFuture {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<()> {
+            Poll::Ready(())
+        }
+    }
+
+    impl Drop for MyFuture {
+        fn drop(&mut self) {
+            println!("task id: {}", task::id());
+        }
+    }
+
+    tokio::spawn(MyFuture).await.unwrap();
+}
```

Comment:
I'm not sure that this test will fail even if a panic happens. You should send a notification back to the main function and assert the success there.